### ISSUE TITLE
General Discussion

### ISSUE THREAD

**Issue by StrykeSlammerII** [EMBTonkin/DAL9001#1]
Question!

You've been writing a custom XML parser, which may make the resulting XML more human-readable and work better with the DRG file format, but in my (mostly non-Java) experience it's easier/better to use [XML Serialization](https://docs.oracle.com/javase/7/docs/api/java/beans/XMLEncoder.html) rather than have to fiddle with the custom parser every time there's a change in the data structures.

I'll see if I can't make a fork for this by Monday, and if the result isn't terrible maybe we can use that as our first merge? :)

( grr no it's a branch we want, not a fork... howto delete a fork? >_<    later: yay found that!)

Ohhhh I see another slight issue: all the Dragon's class variables are wrapped up in the "us" Element. This branch is going to look a bit different internally once I'm done but the getter/setters should all be the same :)

Mm, and your test stuff just dealt with the first Dragon tag found and doesn't address any others it may have found... and doesn't write stuff back out... meh. I'll worry about all that later. Specifically JUST pulling the Dragon tags is gonna be enough work on my end.

....Wait no: How were you going to write the whole XML back out? Oh, that's in DragonTree now. I'll go check how that works, then :)

**Comment by EMBTonkin** [comment]
About serialization: I have never come across this before and it seems like a cool concept.  If my research is accurate, then what it does is take an object and automatically convert it to XML that can be saved?  Very cool.  However, DRG file format is something that a few people came up with together.  Would we be able to to still be able to take in and read out this format?  I'm almost done with the custom XML parsing and it does not seem to be that much trouble.  

(gtg, will write more about your other questions when I get back)
(edit or you know what tomorrow morning it can wait)

**Comment by StrykeSlammerII** [comment]
Yes, serialization converts the object to/from XML automatically. It's pretty cool :)
(It's mostly important in case the relevant data/structures change later, and that may not be an issue for this project.)

Incorporating it into the DRG format may be trickier, but I believe anything inside the <DAL9001> tags does not affect the rest of the format?

Essentially I'll see if I can reverse-engineer a Java class from the existing DRG format. If it fails, we can just ignore my branch.

Your custom parser feels like bad practice to me, but my last big project was about 85% serialization in C# so I'm just biased :D 
"You're wrapping all the data in an XML tree object, and nothing knows what information is actually there??" -me, but you ARE writing the usual get/set methods to expose the data you want, so it's just a really different technique.  I guess if it works, it works... What happens if someone mangles the XML though and the Nodes your methods look for are no longer there?

**Comment by EMBTonkin** [comment]
The reason I am doing what I am doing is that it is a logical next step from where I was.  I might be inadvertently re-inventing the wheel, but I don't know it.  Originally, I read in data, stored it in fields, and then printed it out at the end.  Then when the XML was introduced it was read it, the data extracted and saved in fields, and then the XML object had to be updated right before saving.  It was bad.  

At this point, I still don't see and specific advantage or disadvantage to the serialization vs what I'm doing.  One thing I just though of that I have not pointed out is that the Dragon object is going to hold more that just data from the XML.  The dragon's image object and any other visual things it needs will also be held there, along some other things that I am sure will come up as development continues.  Thus the dragon object is not just a rebuilding of the XML.  It holds a pointer to the XML node and directly edits the XML node so that saving is easier (just print the root).

In the end, this is a fun experimental project, so work on what you think is cool.  I would love to mess around with serialization a little later, I just want to get the framework going first.  

As for mangled XML, the official stance was 'let's hope no one does that.'  xP

**Comment by EMBTonkin** [comment]
Going back to the confusion I see in the original post and hoping to clear some things up:

The Dragon object holds the information of one Dragon.  Hence, only testing the getters and setters on that one specific dragon.  
The DragonTree is the data structure which holds all the dragons, and with hope will be the Model in the MVC.  Maybe it should have been called a Lair :P The internal guts need to be worked out, but this is the object that deals with all the dragons together as one, so it is what you turn to if you need to do things with specific dragons (get them by their ID, get them by whether or not the are exalted, ect) or if you need to do things with all the dragons (load them, save them).  I still am not sure what the internal structure will actually be (in the last one it was a couple lists and some hashmaps).  You can see what I think this object will need to do in the notes at the top of the file.  

Does this help sort things out?  Let me know if you have any more questions, it's best we get on the same page before something horrible happens (eg the program becomes self aware and destroys us all).

**Comment by StrykeSlammerII** [comment]
( I think I got that after reading a bit more of the code but thanks for clarifying ^^ )

Found a possible brick wall on my end:
`//Also also be sure that other people's program's information is preserved, and not eaten.`
That sounds like we'd have to write a custom parser for the XML serialization anyway, to cover any unexpected tags. My bad XD. Since you're already just reading the entire existing XML structure into the Dragon, I think that's gonna end up working best--it just was completely not what I expected :) Thanks for humouring me while I got ahead of myself ^^

OK, so read/write will just deal with the XML, and the getter/setters will modify that--or whatever other backing store for data that doesn't get saved. May be a bit inefficient but shouldn't be so much so that it lags--or if it DOES somehow, we can deal with that later.

A couple new Qs that may need their own Issues:
- Do we have a default constructor for new Dragons without existing XML? (ok according to DragonTree that's a "yes we need it")
- Do we need to add error-handling to the XML-based getter/setters, or do the existing methods error out if they don't find the expected tags?
  (that probably shouldn't be a new Issue, just something I want to check on to know what to expect if someone DOES feed us mangled XML)

...I should probably open a new Issue regarding "what stuff should I work on next"/"what are you doing so I should ignore for now?" XD
You're welcome to close this unless you have any response to this that needs a reply ^^

urp I hit the wrong button--apparently I can edit this but not un-close it? v_v Sorry about that ._.

wait no THERE's the 're-open' button >_<
yay for learning new interface, haha XD

**Comment by EMBTonkin** [comment]
Don't worry about pushing the serialization: it was cool to learn about a new concept, even if it did not end up working out.  Maybe one day when the thing is working fabulously, we can go back and see if we can add it in just for the sake of adding it in!

Second paragraph: Yes to it all

Bullet one: Yup, a constructor for new dragons is something that will need to happen, since we can hardly expect everyone to already have all their dragon's data in an XML file.  exactly HOW to go about doing this is something I have not yet put much thought into.  One way I have thought would be having DragonTree have an addDragon function that creates a Node with the data and then created the Dragon object for that node.  It's a but clunky, do you have any ideas?

Bullet two: We don't need error handlers, everything is PERFECT!  Actually we should probably add those in at some point, as the program would probably crash from some sort of Null pointer if the XML was not properly formatted.  Good call.

I've opened said new issue and stated what I am working on.  If we can get pulling and pushing to work properly then if we just update the code every time we do something there will hopefully not be too much overlap.

**Comment by StrykeSlammerII** [comment]
re: adding/new Dragons
I made a default constructor for Dragon that creates a new XML structure with mostly-empty tags, and default constructor for DragonTree that creates the tags that wrap around the `<Dragon>` stuff.
Also made DragonTree.AddDragon( Dragon newDragon ) and added test functionality to the main() in DragonTree to make a new (blank) lair with a single new (blank) dragon, and write that out.
Then it does the old read-write of the demo .drg file.

In practical use though, there's gonna be a button or something that brings up a new Dragon and has the user populate stuff, and then add that to the DragonTree. 

Once I can get commits to update the repository here (or figure out whatever command I'm supposed to be using), then you should have all that too :)

**Comment by EMBTonkin** [comment]
Ooooh, those sound like exactly the sort of functions we need!  I cannot wait to see them!

Practically, yes there is going to be a button that brings up a window or something where you enter data for the new dragon, and then creates the dragon with that data.

**Comment by EMBTonkin** [comment]
I was away for a few days (just so you don't think I was being lazy), I'll look over your commit tomorrow morning.  Also merges count as a separate commit which is where the 'merge branch 'master' thing' came from.

**Comment by EMBTonkin** [comment]
I made a mockup of what the UI will look like.  
(Please ignore the bad GIMP and weird colors.  )
What do you think of it?  The red box will have all the control buttons for adding, exalting, saving, editing, and other advanced stuff we come up with.  Cyan box is where data will be displayed, and here the black boxes represent a line of colored boxes (you know the usual color range display).  Olive section is for the actual Dragon family tree.  
<img width="1394" alt="dal mockup" src="https://cloud.githubusercontent.com/assets/8117715/12870429/5d4d7c62-cd12-11e5-9b78-e84c9e250b6b.png">

This is different than the old one which had the stats crammed in the sidebar below the buttons.  This will not work once the color expansion comes out and color ranges are longer, hence the new header section.  Let me know what you think.  

Also, why does Java have 4 layout managers, none of which do what I want?

**Comment by StrykeSlammerII** [comment]
Looks like a decent layout for starters.
My main (rhetorical-for-now) Qs: 1) will the red section be TOO big? Or will it end up just a single column of buttons?
2) Can we make the red and blue sections expandable (like they were frames and you can drag the interior border up-down or left-right to take more room) ?

Lol  I haven't looked at Java's layout managers in ages, as previously noted--but of course they won't have the one you need, what did you expect? haha ;)
[ ok time for me to do some work for the weekend -_- ]

**Comment by EMBTonkin** [comment]
1) The red section will be a single column of buttons, so I'll make it as wide as the widest button.
2) That is a super awesome idea! and it's probably impossible!  I'll look into that later once I get other aspects of the UI working.
